microsoft/pai
# Use Postgresql as an alternative storage for job history

#2850 

How to use

The following service config should be set to enable internal storage, postgresql, and set rest-server to read job history from database.

```yaml
cluster:
  common:
    job-history: "true"
    job-attempt-storage-engine: "db"

internal-storage:
  root-path: /paiInternal
  enable: true

postgresql:
  enable: true
```

## What will happen if the database crashes

In such case, the health check of job attempt api will not pass. User won't be able to click it.


![image](https://user-images.githubusercontent.com/7499023/72954917-1651a800-3dd5-11ea-9665-8c8ae817b22e.png)

In backend side, `fluentd` will try to send chunk to postgresql forever. Once the postgresql db is restarted, the job attempt history will be recovered.

## Test

0. Check if the internal-storage and postgresql db can be deployed 

- [x] Check if the job history works normally
   - Submit a special job, which will fail after 60 seconds: `sleep 60s; echo failed at $(date); exit 233`, and set a retry numbers for it.
   - Expect: all the job retry history will be recorded normally.

- [x] Check the behavior when the database is down
   - Delete the database service by `./paictl.py service stop -m postgresql`
   - Submit a job with retries, now the user will not be able to see the retry history
   - Start the database service by `./paictl.py service start -m  postgresql`
   - Expect: all the job retry history will be recorded normally after a few minutes

- [x] Kill the database process to check if it can recover
   - find the database process in docker, kill it.
   - check if it can recover itself
   - check if the job history server can become normal within a few minutes

- [x] Bulk test: submit a job with 1000 instances, stop the postgres db from time to time.
